### PR TITLE
[HttpKernel] Deprecate passing objects as URI attributes to the ESI and SSI renderers

### PR DESCRIPTION
**Pull request by jakzal**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #15056 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by fabpot** [comment]
Can you update the CHANGELOG and UPGRADE files?

File: src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php [hunk 1]
```diff
@@ -64,6 +64,10 @@ public function __construct(SurrogateInterface $surrogate = null, FragmentRender
     public function render($uri, Request $request, array $options = array())
     {
         if (!$this->surrogate || !$this->surrogate->hasSurrogateCapability($request)) {
+            if ($uri instanceof ControllerReference && $this->containsNonScalars($uri->attributes)) {
+                @trigger_error('Passing objects as part of URI attributes to the ESI and SSI rendering strategies is deprecated. Use another rendering strategy or use scalar values.', E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php]: you should mention that this is deprecated as of 3.1 and that this possibility will be removed in 4.0.

File: src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php [hunk 2]
```diff
@@ -92,4 +96,17 @@ private function generateSignedFragmentUri($uri, Request $request)
 
         return substr($fragmentUri, strlen($request->getSchemeAndHttpHost()));
     }
+
+    private function containsNonScalars(array $values)
+    {
+        foreach ($values as $key => $value) {
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php]: `$key` can be removed

File: src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php [hunk 3]
```diff
@@ -92,4 +96,17 @@ private function generateSignedFragmentUri($uri, Request $request)
...
+            if (is_array($value)) {
+                return $this->containsNonScalars($value);
+            } elseif (!is_scalar($value) && null !== $value) {
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php]: Why is `null` special?

Comment (jakzal) [inline src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php]: [`is_scalar()` returns `false` with `null`](https://3v4l.org/GEBkU). We want to support nulls hence the second part of the condition.

Comment (stloyd) [inline src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php]: Should be just `if` not `elseif` as you call `return` in other `if`.

**Comment by jakzal** [comment]
Addressed the comments.

File: CHANGELOG-3.1.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+This changelog references the relevant changes (bug and security fixes) done
+in 3.1 minor versions.
+
+To get the diff for a specific change, go to https://github.com/symfony/symfony/commit/XXX where XXX is the change hash
+To get the diff between two versions, go to https://github.com/symfony/symfony/compare/v3.1.0...v3.1.1
+
+* 3.1.0
+
+ * bug #17611 [HttpKernel] Deprecate passing objects as URI attributes to the ESI and SSI renderers
```
Comment (fabpot) [inline CHANGELOG-3.1.md]: This file is auto-generated. I meant the CHANGELOG file for the HttpKernel Component instead. Sorry for the confusion.

**Comment by fabpot** [comment]
:+1:

File: src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php (hunk 3 above)
Comment (jakzal) [inline src/Symfony/Component/HttpKernel/Fragment/AbstractSurrogateFragmentRenderer.php]: @stloyd actually, the code wasn't looking deep enough. I updated the test and the condition. Now the elseif is required.

**Comment by fabpot** [comment]
Thank you @jakzal.

File: UPGRADE-3.1.md [hunk 5]
```diff
@@ -16,6 +16,13 @@ Form
  * The `choices_as_values` option of the `ChoiceType` has been deprecated and
    will be removed in Symfony 4.0.
 
+HttpKernel
+----------
+
+ * Passing objects as URI attributes to the ESI and SSI renderers has been
+   deprecated and will be removed in Symfony 4.0. The inline fragment
+   renderer should be used with object attributes.
```
Comment (xabbuh) [inline UPGRADE-3.1.md]: missing entry about the removal in the `UPGRADE-4.0.md` file

Comment (jakzal) [inline UPGRADE-3.1.md]: It's not removed yet :)

**Pull request merged**